File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.py
from collections import deque
import math

def solution(players, m, k):
    answer = 0
    
    queue = deque()
        
    for i in range(24):
        while queue:
            if i-queue[0] < k:
                break
                
            queue.popleft()
            
        totalServer = len(queue)
        need = 0 if players[i] < m else players[i]//m
        
        print(totalServer, need)
        
        if need > totalServer:
            for j in range((need-totalServer)):
                queue.append(i)
                answer += 1
    
    return answer

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.*;

class Solution {
    public int solution(int[] players, int m, int k) {
        int answer = 0;
        int totalServer = 0;
        int need = 0;
        
        // 작동 중인 서버
        Deque<Integer> queue = new ArrayDeque<>();
        
        for(int i=0; i<players.length; ++i){
            // 작동 시간 끝난 서버 큐에서 제거
            while(!queue.isEmpty()){
                if(i-queue.peekFirst() < k) break;
                queue.pollFirst();                
            }
            
            totalServer = queue.size();
            need = (int) Math.ceil(players[i]/m);
            
            // 서버 증설이 필요한 경우 큐에 추가
            if (need > totalServer) {
                for(int j=0; j<need-totalServer; ++j){
                    queue.addLast(i);
                    answer++;
                }
            }   
        }
         
        return answer;
    }
}